subsys/bluetooth/mesh/shell/shell_utils.c

```diff
@@ -37,70 +37,22 @@ static void model_instances_get(uint16_t id, struct shell_model_instance *arr, u
 	}
 }
 
-static uint8_t str2u8(const char *str)
+struct sensor_value shell_model_strtosensorval(const char *str, int *err)
 {
-	if (isdigit((unsigned char)str[0])) {
-		return strtoul(str, NULL, 0);
-	}
-
-	return (!strcmp(str, "on") || !strcmp(str, "enable"));
-}
-
-bool shell_model_str2bool(const char *str)
-{
-	return str2u8(str);
-}
-
-
-int shell_model_str2sensorval(const char *str, struct sensor_value *out)
-{
-	int32_t sign = 1;
-	char buf[18]; /* 10 digits + decimal point + 6 digits + terminator */
-
-	if (*str == '-') {
-		sign = -1;
-		str++;
-	}
+	int temp_err = 0;
+	struct sensor_value out;
 
-	if (strlen(str) > 17 || strlen(str) == 0) {
-		return -EINVAL;
-	}
-
-	strcpy(buf, str);
-
-	char *dp = strchr(buf, '.');
-
-	if (dp != NULL) {
-		*dp = 0;
-		dp++;
-	}
+	double val = shell_model_strtodbl(str, &temp_err);
 
-	errno = 0;
-	uint32_t val = strtoul(buf, NULL, 10);
-
-	if (errno || (sign == 1 && val > INT32_MAX) || (sign == -1 && val > INT32_MAX + 1ul)) {
-		return -EINVAL;
-	}
-	out->val1 = sign * val;
-	if (dp == NULL || *dp == 0) {
-		out->val2 = 0;
-		return 0;
+	if (temp_err) {
+		*err = temp_err;
+		return out;
 	}
 
-	int32_t factor = 100000;
-
-	val = 0;
+	out.val1 = (int)val;
+	out.val2 = (val - (double)out.val1) * 1000000;
```

Comment:
Nitpick: The `(double)` cast on this line is implicit in C if `val` is a `double`